File: teamapps-client/projector-document-viewer/src/main/java/org/teamapps/projector/component/documentviewer/DocumentViewer.java
/*-
 * ========================LICENSE_START=================================
 * TeamApps
 * ---
 * Copyright (C) 2014 - 2022 TeamApps.org
 * ---
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * =========================LICENSE_END==================================
 */
package org.teamapps.projector.component.documentviewer;

import org.teamapps.projector.annotation.ClientObjectLibrary;
import org.teamapps.projector.component.AbstractComponent;
import org.teamapps.projector.component.ComponentConfig;
import org.teamapps.projector.format.Border;
import org.teamapps.projector.format.BoxShadow;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

@ClientObjectLibrary(DocumentViewerLibrary.class)
public class DocumentViewer extends AbstractComponent implements DtoDocumentViewerEventHandler{

	private final DtoDocumentViewerClientObjectChannel clientObjectChannel = new DtoDocumentViewerClientObjectChannel(getClientObjectChannel());

	private List<String> pageUrls = new ArrayList<>();
	private PageDisplayMode displayMode = PageDisplayMode.FIT_WIDTH;
	private float zoomFactor = 1;
	private Border pageBorder;
	private BoxShadow pageShadow;
	private int padding;
	private int pageSpacing = 5;

	public DocumentViewer() {
	}

	public DocumentViewer(List<String> pageUrls) {
		this.pageUrls.addAll(pageUrls);
	}

	@Override
	public ComponentConfig createConfig() {
		DtoDocumentViewer documentViewer = new DtoDocumentViewer();
		mapAbstractUiComponentProperties(documentViewer);
		documentViewer.setPageUrls(pageUrls);
		documentViewer.setDisplayMode(displayMode);
		documentViewer.setZoomFactor(zoomFactor);
		documentViewer.setPageBorder(pageBorder != null ? pageBorder.createUiBorder() : null);
		documentViewer.setPageShadow(pageShadow != null ? pageShadow.createUiShadow() : null);
		documentViewer.setPadding(padding);
		documentViewer.setPageSpacing(pageSpacing);
		return documentViewer;
	}

	public List<String> getPageUrls() {
		return pageUrls;
	}

	public void setPageUrls(List<String> pageUrls) {
		this.pageUrls.clear();
		this.pageUrls.addAll(pageUrls);
		clientObjectChannel.setPageUrls(this.pageUrls);
	}

	public void setPageUrls(String... pageUrls) {
		setPageUrls(Arrays.asList(pageUrls));
	}

	public void addPageUrl(String pageUrl) {
		this.pageUrls.add(pageUrl);
		clientObjectChannel.setPageUrls(this.pageUrls);
	}

	public PageDisplayMode getDisplayMode() {
		return displayMode;
	}

	public void setDisplayMode(PageDisplayMode displayMode) {
		this.displayMode = displayMode;
		clientObjectChannel.setDisplayMode(displayMode, zoomFactor);
	}

	public float getZoomFactor() {
		return zoomFactor;
	}

	public void setZoomFactor(float zoomFactor) {
		this.zoomFactor = zoomFactor;
		clientObjectChannel.setZoomFactor(zoomFactor);
	}

	public Border getPageBorder() {
		return pageBorder;
	}

	public void setPageBorder(Border pageBorder) {
		this.pageBorder = pageBorder;
		clientObjectChannel.setPageBorder(pageBorder != null ? pageBorder.createUiBorder(): null);
	}

	public BoxShadow getPageShadow() {
		return pageShadow;
	}

	public void setPageShadow(BoxShadow pageShadow) {
		this.pageShadow = pageShadow;
		clientObjectChannel.setPageShadow(pageShadow != null ? pageShadow.createUiShadow(): null);
	}

	public int getPadding() {
		return padding;
	}

	public void setPadding(int padding) {
		this.padding = padding;
		clientObjectChannel.setPaddding(padding);
	}

	public int getPageSpacing() {
		return pageSpacing;
	}

	public void setPageSpacing(int pageSpacing) {
		this.pageSpacing = pageSpacing;
		clientObjectChannel.setPageSpacing(pageSpacing);
	}

}

File: client/projector-document-viewer/src/main/ts/DocumentViewer.ts
/*-
 * ========================LICENSE_START=================================
 * TeamApps
 * ---
 * Copyright (C) 2014 - 2022 TeamApps.org
 * ---
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * =========================LICENSE_END==================================
 */

import {
	AbstractComponent,
	applyCss, createBorderCssString, createShadowCssString,
	type DtoBorder,
	type DtoBoxShadow,
	generateUUID,
	parseHtml
} from "projector-client-object-api";
import {type DtoDocumentViewer, type DtoDocumentViewerCommandHandler, type PageDisplayMode, PageDisplayModes} from "./generated";
import {enableScrollViaDragAndDrop} from "./scroll-dnd";

interface Page {
	$img: HTMLElement;
	naturalWidth: number;
	naturalHeight: number;
}

export class DocumentViewer extends AbstractComponent<DtoDocumentViewer> implements DtoDocumentViewerCommandHandler {

	private $componentWrapper: HTMLElement;
	private $pagesContainerWrapper: HTMLElement;
	private $pagesContainer: HTMLElement;
	private zoomFactor: number;
	private displayMode: PageDisplayMode;
	private pages: Page[] = [];

	private uuidClass: string;
	private $styleTag: HTMLElement;

	constructor(config: DtoDocumentViewer) {
		super(config);

		this.uuidClass = `DocumentViewer-${generateUUID()}`;

		this.$componentWrapper = parseHtml(`<div class="DocumentViewer ${this.uuidClass}">  
		    <style></style>
			<div class="toolbar-container"></div>  
			<div class="pages-container-wrapper">
			    <div class="pages-container"></div>
		    </div>
	    </div>`);

		this.$pagesContainerWrapper = this.$componentWrapper.querySelector<HTMLElement>(":scope .pages-container-wrapper")!;
		this.$styleTag = this.$componentWrapper.querySelector<HTMLElement>(":scope style")!;
		this.$pagesContainer = this.$componentWrapper.querySelector<HTMLElement>(':scope .pages-container')!;
		enableScrollViaDragAndDrop(this.$pagesContainerWrapper);

		this.zoomFactor = config.zoomFactor;
		this.displayMode = config.displayMode;

		if (config.pageUrls) {
			this.setPageUrls(config.pageUrls);
		}

		this.updateStyles()
	}

	public setPageUrls(pageUrls: string[]) {
		this.$pagesContainer.innerHTML = '';
		pageUrls.forEach((pageUrl) => {
			const img = new Image();
			let page: Page = {
				$img: img,
				naturalWidth: 0,
				naturalHeight: 0
			};

			img.onload = () => {
				page.naturalWidth = img.naturalWidth;
				page.naturalHeight = img.naturalHeight;
				this.updateImageSizes();
			};

			img.src = pageUrl;
			img.classList.add("page");

			this.pages.push(page);
			this.$pagesContainer.append(img);
		});
		this.updateImageSizes();
	}

	public setDisplayMode(displayMode: PageDisplayMode, zoomFactor: number) {
		this.displayMode = displayMode;
		this.zoomFactor = zoomFactor;
		this.updateImageSizes();
	}

	public setZoomFactor(zoomFactor: number) {
		this.zoomFactor = zoomFactor;
		this.updateImageSizes();
	}

	private updateImageSizes() {
		let viewPortWidth = this.$pagesContainerWrapper.offsetWidth - 2 * this.config.padding;
		let viewPortHeight = this.$pagesContainerWrapper.offsetHeight - 2 * this.config.padding;
		let viewPortAspectRatio = viewPortWidth / viewPortHeight;

		this.pages.forEach((p) => {
			let imageAspectRatio = p.naturalWidth / p.naturalHeight;
			console.debug("image: " + p.naturalWidth + "/" + p.naturalHeight + " = " + imageAspectRatio);
			console.debug("viewport: " + viewPortWidth + "/" + viewPortHeight + " = " + viewPortAspectRatio);
			if (this.displayMode === PageDisplayModes.FIT_WIDTH) {
				applyCss(p.$img, {
					width: Math.floor(viewPortWidth * this.zoomFactor) + "px",
					height: "auto"
				});
			} else if (this.displayMode === PageDisplayModes.FIT_HEIGHT) {
				applyCss(p.$img, {
					width: "auto",
					height: Math.floor(viewPortHeight * this.zoomFactor) + "px"
				});
			} else if (this.displayMode === PageDisplayModes.FIT_SIZE) {
				if (imageAspectRatio > viewPortAspectRatio) {
					applyCss(p.$img, {
						width: Math.floor(viewPortWidth * this.zoomFactor) + "px",
						height: "auto"
					});
				} else {
					applyCss(p.$img, {
						width: "auto",
						height: Math.floor(viewPortHeight * this.zoomFactor) + "px"
					});
				}
			} else if (this.displayMode === PageDisplayModes.COVER) {
				if (imageAspectRatio < viewPortAspectRatio) {
					applyCss(p.$img, {
						width: Math.floor(viewPortWidth * this.zoomFactor) + "px",
						height: "auto"
					});
				} else {
					applyCss(p.$img, {
						width: "auto",
						height: Math.floor(viewPortHeight * this.zoomFactor) + "px"
					});
				}
			} else {
				applyCss(p.$img, {
					width: (p.naturalWidth * this.zoomFactor) + "px",
					height: "auto"
				});
			}
		});
	}

	onResize(): void {
		this.updateImageSizes();
	}

	public doGetMainElement(): HTMLElement {
		return this.$componentWrapper;
	}

	setPageBorder(pageBorder: DtoBorder): void {
		this.config.pageBorder = pageBorder;
		this.updateStyles();
	}

	setPaddding(padding: number): void {
		this.config.padding = padding;
		this.updateStyles();
	}

	setPageSpacing(pageSpacing: number): void {
		this.config.pageSpacing = pageSpacing;
		this.updateStyles();
	}

	setPageShadow(pageShadow: DtoBoxShadow): void {
		this.config.pageShadow = pageShadow;
		this.updateStyles();
	}

	private updateStyles() {
		this.$styleTag.innerHTML = '';
		this.$styleTag.innerText = `
		.${this.uuidClass} .page {
            ${createBorderCssString(this.config.pageBorder)}
            ${createShadowCssString(this.config.pageShadow)}
        }
        .${this.uuidClass} .page:not(:last-child) {
            margin-bottom: ${this.config.pageSpacing}px !important;
        }`;
		this.$pagesContainer.style.padding = this.config.padding + "px";
	}
}


